data-pipeline/src/concentrator/aggregation.rs

```diff
@@ -0,0 +1,497 @@
+// Copyright 2024-Present Datadog, Inc. https://www.datadoghq.com/
+// SPDX-License-Identifier: Apache-2.0
+//! This module implement the logic for stats aggregation into time buckets and stats group.
+//! This includes the aggregation key to group spans together and the computation of stats from a
+//! span.
+use datadog_trace_protobuf::pb;
+use datadog_trace_utils::trace_utils::has_top_level;
+use ddcommon::tag::Tag;
+use std::collections::HashMap;
+
+const TAG_STATUS_CODE: &str = "http.status_code";
+const TAG_SYNTHETICS: &str = "synthetics";
+const TAG_SPANKIND: &str = "span.kind";
+const TAG_ORIGIN: &str = "_dd.origin";
+
+/// This struct represent the key used to group spans together to compute stats.
+#[derive(Debug, Hash, PartialEq, Eq, Clone, Default)]
+pub struct AggregationKey {
+    pub resource_name: String,
+    pub service_name: String,
+    pub operation_name: String,
+    pub span_type: String,
+    pub span_kind: String,
+    pub http_status_code: u32,
+    pub is_synthetics_request: bool,
+    pub peer_tags: Vec<Tag>,
+    pub is_trace_root: bool,
+}
+
+impl AggregationKey {
+    /// Return an AggregationKey matching the given span.
+    ///
+    /// If `peer_tags_aggregation` is true then the peer tags of the span will be included in the
+    /// key. The peer tags are computed based on the list of `peer_tag_keys`.
+    pub fn from_span(
+        span: &pb::Span,
+        peer_tags_aggregation: bool,
+        peer_tag_keys: &[String],
+    ) -> Self {
+        let span_kind = span
+            .meta
+            .get(TAG_SPANKIND)
+            .map(|s| s.to_string())
+            .unwrap_or_default();
+        let peer_tags = if peer_tags_aggregation && client_or_producer(&span_kind) {
+            get_peer_tags(span, peer_tag_keys)
+        } else {
+            vec![]
+        };
+        Self {
+            resource_name: span.resource.clone(),
+            service_name: span.service.clone(),
+            operation_name: span.name.clone(),
+            span_type: span.r#type.clone(),
```

Comment:
It feels a bit costly to _always_ (at least for relevant spans) construct this whole Key by cloning a number of things, only to use it to look up things, and eventually throw it away. Something to keep in mind when we start measuring this code.

Reply:
I agree, I think https://docs.rs/hashbrown/latest/hashbrown/hash_map/struct.HashMap.html#method.raw_entry_mut would be the way to go. I added APMSP-1331 to track this.

Reply:
Great